xlators/mgmt/glusterd/src/glusterd-utils.c

```diff
@@ -2112,6 +2113,8 @@ glusterd_volume_start_glusterfs(glusterd_volinfo_t *volinfo,
         ret = -1;
         goto out;
     }
+
+    errno_tmp = errno;
```

Comment:
In a race condition, the `pmap_assign_port()` might succeed but with the same port number for two different brick...so in that case, the bind() will still fail and set the `errno` to the appropriate value.....though I am not sure if `sys_close()` could modify the `errno` or not, because in my testing over gdb it didn't.

But, yes to be on a safer side we can do the same inside the `pmap_port_isfree()` and return it as a negative value. Will check how the same can be acheived.

Reply:
I tried, looking for a way to set it inside the `pmap_port_isfree()` and then pass it to the caller, but that doesn't seem to work out well. 
If we return a negative port number instead of a valid port number or 0 (in case of max port reached), then the other parts of the code will be affected and will not work as expected such as in this function itself, we are assigning the returned port number to `brickinfo->port` and then executing the `runner_run()` with those details for the brick and the will not work correctly.

So, I think the current implementation is fine because the only possibility of the `errno` being changed is if `sys_lock()` fails, else the value will be retained....and even if `sys_lock()` fails then we would be logging the failure for it.

Hence, even if we continue with this approach we won't be breaking anything or causing any issue.

Reply:
> In a race condition, the `pmap_assign_port()` might succeed but with the same port number for two different brick...

Without patch #1670 I don't think it's possible to assign the same port to two bricks because as soon as one port is assigned, it gets reserved and cannot be issues to another brick until the first one releases it. If this problem happens after patch #1670, then I think what needs to be fixed is the other patch.

> so in that case, the bind() will still fail and set the `errno` to the appropriate value.....though I am not sure if `sys_close()` could modify the `errno` or not, because in my testing over gdb it didn't.

`pmap_registry_alloc()` in the current code or `pmap_port_alloc()` after PR #1670 don't return a port that fails `bind()`, because a new port is tried. So I don't see what `errno` you want to use. On success `errno` may not be updated, so the "returned" errno may refer to a previous failed attempt that is stale.

Reply:
> I tried, looking for a way to set it inside the `pmap_port_isfree()` and then pass it to the caller, but that doesn't seem to work out well.
> If we return a negative port number instead of a valid port number or 0 (in case of max port reached), then the other parts of the code will be affected and will not work as expected such as in this function itself, we are assigning the returned port number to `brickinfo->port` and then executing the `runner_run()` with those details for the brick and the will not work correctly.

Of course, you need to correctly handle a negative value (note that both 0 and a negative number are an invalid port number).

> So, I think the current implementation is fine because the only possibility of the `errno` being changed is if `sys_lock()` fails, else the value will be retained....and even if `sys_lock()` fails then we would be logging the failure for it.

I strongly disagree. You cannot use an `errno` value after a successful call, even if we assume that `errno` lives enough to reach this point without modification, which I wouldn't bet on it.

Note that before this patch, line 2117 is only executed if `port` is a positive number, so it has succeeded, which means that 'errno' contains garbage and cannot be used.